src/GUI/UI/Houses/Training.cpp

```diff
@@ -25,198 +25,176 @@ IndexedArray<int, HOUSE_TRAINING_HALL_EMERALD_ISLE, HOUSE_TRAINING_HALL_STONE_CI
-                                        LSTR_FMT_TRAIN_LEVEL_D_FOR_D_GOLD,
-                                        pParty->activeCharacter().uLevel + 1,
-                                        pPrice);
-                                pShopOptions[index] = shop_option_str_container.c_str();
-                            }
-                        }
-                        all_text_height += pFontArrus->CalcTextHeight(pShopOptions[index], training_dialog_window.uFrameWidth, 0);
-                        ++index;
-                    }
-                }
-                v49 =
-                    (2 * (87 - (SIDE_TEXT_BOX_BODY_TEXT_HEIGHT - all_text_height) / 2) - all_text_height) /
-                    2 -
-                    (SIDE_TEXT_BOX_BODY_TEXT_HEIGHT - all_text_height) / 2 / 2 + SIDE_TEXT_BOX_BODY_TEXT_OFFSET;
-                if (pDialogueWindow->pStartingPosActiveItem <
-                    pDialogueWindow->pStartingPosActiveItem +
-                    pDialogueWindow->pNumPresenceButton) {
-                    index = 0;
-                    for (int i = pDialogueWindow->pStartingPosActiveItem;
-                        i < pDialogueWindow->pStartingPosActiveItem +
-                        pDialogueWindow->pNumPresenceButton;
-                        ++i) {
-                        pButton = pDialogueWindow->GetControl(i);
-                        pButton->uY = (SIDE_TEXT_BOX_BODY_TEXT_HEIGHT - all_text_height) / 2 + v49;
-                        pTextHeight = pFontArrus->CalcTextHeight(
-                            pShopOptions[index],
-                            training_dialog_window.uFrameWidth, 0);
-                        pButton->uHeight = pTextHeight;
-                        pButton->uW = pTextHeight + pButton->uY - 1 + 6;
-                        v49 = pButton->uW;
-                        Color pTextColor = colorTable.Jonquil;
-                        if (pDialogueWindow->pCurrentPosActiveItem != i)
-                            pTextColor = colorTable.White;
-                        training_dialog_window.DrawTitleText(
-                            pFontArrus, 0, pButton->uY, pTextColor,
-                            pShopOptions[index], 3);
-                        ++index;
+        int pPrice = PriceCalculator::trainingCostForPlayer(&pParty->activeCharacter(), buildingTable[wData.val - 1]);
+        uint64_t expForNextLevel = 1000ull * pParty->activeCharacter().uLevel * (pParty->activeCharacter().uLevel + 1) / 2;
+        int index = 0;
+        int all_text_height = 0;
+
+        pShopOptions[0] = "";
+        pShopOptions[1] = localization->GetString(LSTR_LEARN_SKILLS);
+        for (int i = pDialogueWindow->pStartingPosActiveItem; i < pDialogueWindow->pNumPresenceButton + pDialogueWindow->pStartingPosActiveItem; ++i) {
+            if (pDialogueWindow->GetControl(i)->msg_param == DIALOGUE_TRAINING_HALL_TRAIN) {
+                static std::string shop_option_str_container;
```

Comment:
Why static? Let's add a TODO at least to make it non-static. Should be doable by replacing `char *` with `std::string` in `pShopOptions`.